### PR TITLE
parse attributes

### PR DESCRIPTION
**Pull request by GrayHatter**

### CODE REVIEW

**Comment by GrayHatter** [comment]
see #2

**Comment by GrayHatter** [comment]
see #6

File: include/scalars.h [hunk 1]
```diff
@@ -20,17 +21,24 @@ enum scalar_type {
...
+
+	SCALAR_BOOL,
+	SCALAR_CHAR,
+	SCALAR_VOID,
 };
 
 struct sui_scalar {
 	enum scalar_type type;
 	union {
+		bool B;
```
Comment (SirCmpwn) [inline include/scalars.h]: Name this bval

File: include/scalars.h [hunk 2]
```diff
@@ -20,17 +21,24 @@ enum scalar_type {
...
 		double fval;
 		long ival;
 		unsigned long uval;
 		struct {
 			uint8_t a, b, g, r;
 		};
+		char *c;
+		void *p;
```
Comment (SirCmpwn) [inline include/scalars.h]: `str` and `data`

File: sui/parse_attrib.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+#include "node.h"
+#include "subparser.h"
+#include "parser.h"
+#include "scalars.h"
+
+enum string_type {
+	STRING_LITERAL,
+	STRING_DOUBLE_QUOTE,
+	STRING_SINGLE_QUOTE
+};
```
Comment (SirCmpwn) [inline sui/parse_attrib.c]: Any reason you didn't reuse the string parser we already have?

File: sui/parse_attrib.c [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+void parse_attrib(struct parser_state *pstate, uint32_t ch) {
+	struct subparser_state *subparser = list_peek(pstate->parsers);
+	struct attrib_state *state = subparser->state;
+
+	switch(state->phase) {
+	case PHASE_UNKNOWN: {
+		state->phase = PHASE_ATTRIB_KEY;
+		// Intentional fall through
+	}
```
Comment (SirCmpwn) [inline sui/parse_attrib.c]: Don't add these braces where they're not necessary. Also, just `/* fallthrough */` will do

File: test/sui/sui_parse.c [hunk 5]
```diff
@@ -112,11 +113,56 @@ static int test_indentation_errors() {
 	return 0;
 }
 
+static int test_attrib_types() {
+	errors_t *err = NULL;
+	struct sui_node *node = sui_parse("test\n"
+	                               "    some_node blerg=rip\n"
+	                               "    other_node blerg = something\n"
+	                               "    other_node2 blerg = something else", &err);
+
```
Comment (SirCmpwn) [inline test/sui/sui_parse.c]: Just indent these once greater than the first line and use tabs. Also, recommended sui style will not put a space between key=value, please follow that here

File: include/scalars.h [hunk 6]
```diff
@@ -5,6 +5,7 @@
 #include <stdbool.h>
 
 enum scalar_type {
+	SCALAR_UNKNOWN,
```
Comment (SirCmpwn) [inline include/scalars.h]: Call this `SCALAR_EMPTY`?

File: include/scalars.h [hunk 7]
```diff
@@ -20,17 +21,24 @@ enum scalar_type {
 	SCALAR_HZ,
 	SCALAR_KHZ,
 	SCALAR_COLOR,
+
```
Comment (SirCmpwn) [inline include/scalars.h]: Why the space?

File: test/sui/sui_parse.c (hunk 5 above)
Comment (GrayHatter) [inline test/sui/sui_parse.c]: I wrote this test to make sure we can support both styles. the first is recommended, the latter is something we want to support for obvious reasons, would you like me to add a comment, change the text of the key value to make it obvious this is the wrong style? or delete the test for this use case.

File: sui/parse_attrib.c (hunk 4 above)
Comment (GrayHatter) [inline sui/parse_attrib.c]: done

File: sui/parse_attrib.c (hunk 3 above)
Comment (GrayHatter) [inline sui/parse_attrib.c]: Per our previous IRC conversation; the string parser is less strict than we want to be to parse attribute names/keys. Unless I misunderstood?

2nd, the string parser works while attributes are just "string" = "string" but it's my understanding we want to parse some options differently like `key = true` should return a bool scalar?

File: include/scalars.h (hunk 2 above)
Comment (GrayHatter) [inline include/scalars.h]: done

File: include/scalars.h (hunk 1 above)
Comment (GrayHatter) [inline include/scalars.h]: done

File: include/scalars.h (hunk 7 above)
Comment (GrayHatter) [inline include/scalars.h]: to make this comment easier, I wanted to make sure you could comment on how you wanted it sorted.

I'll just remove the space

File: include/scalars.h (hunk 6 above)
Comment (GrayHatter) [inline include/scalars.h]: done

File: sui/parse_attrib.c (hunk 3 above)
Comment (SirCmpwn) [inline sui/parse_attrib.c]: I would just use the same string parser and validate the string in a second step.

Comment (SirCmpwn) [inline sui/parse_attrib.c]: Leave types aside for now.

Comment (SirCmpwn) [inline sui/parse_attrib.c]: Even when we do them, we'll parse it as a string and hand the string to an entirely separate parser to obtain a record of the correct type.

File: test/sui/sui_parse.c (hunk 5 above)
Comment (SirCmpwn) [inline test/sui/sui_parse.c]: It looks like you don't have any tests that use the recommended style. I would just test it with spaces in one of these lines, not all of them.

**Pull request merged**